test/e2e/storage/testsuites/fsgroupchangepolicy.go

```diff
@@ -0,0 +1,278 @@
+	skipVolTypePatterns(pattern, driver, testpatterns.NewVolTypeMap(testpatterns.CSIInlineVolume, testpatterns.GenericEphemeralVolume))
+}
+
+func (s *fsGroupChangePolicyTestSuite) DefineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	type local struct {
+		config        *PerTestConfig
+		driverCleanup func()
+		driver        TestDriver
+		resource      *VolumeResource
+	}
+	var l local
+	ginkgo.BeforeEach(func() {
+		dInfo := driver.GetDriverInfo()
+		if !dInfo.Capabilities[CapFsGroup] {
+			e2eskipper.Skipf("Driver %q does not support FsGroup - skipping", dInfo.Name)
+		}
+
+		if pattern.VolMode == v1.PersistentVolumeBlock {
+			e2eskipper.Skipf("Test does not support non-filesystem volume mode - skipping")
+		}
+
+		if pattern.VolType != testpatterns.DynamicPV {
+			e2eskipper.Skipf("Suite %q does not support %v", s.tsInfo.Name, pattern.VolType)
+		}
+
+		_, ok := driver.(DynamicPVTestDriver)
+		if !ok {
+			e2eskipper.Skipf("Driver %s doesn't support %v -- skipping", dInfo.Name, pattern.VolType)
+		}
+
+		if framework.NodeOSDistroIs("windows") {
+			e2eskipper.Skipf("FSGroupPolicy is only applied on linux nodes -- skipping")
+		}
+	})
+	// This intentionally comes after checking the preconditions because it
+	// registers its own BeforeEach which creates the namespace. Beware that it
+	// also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
+	f := framework.NewDefaultFramework("fsgroupchangepolicy")
+
+	init := func() {
+		l = local{}
+		l.driver = driver
+		l.config, l.driverCleanup = driver.PrepareTest(f)
+		testVolumeSizeRange := s.GetTestSuiteInfo().SupportedSizeRange
+		l.resource = CreateVolumeResource(l.driver, l.config, pattern, testVolumeSizeRange)
+	}
+
+	cleanup := func() {
+		var errs []error
+		if l.resource != nil {
+			if err := l.resource.CleanupResource(); err != nil {
+				errs = append(errs, err)
+			}
+			l.resource = nil
+		}
+
+		errs = append(errs, tryFunc(l.driverCleanup))
+		l.driverCleanup = nil
+		framework.ExpectNoError(errors.NewAggregate(errs), "while cleanup resource")
+	}
+
+	ginkgo.It("pod fsgroup change policy (Always) should always change the volume permission to match fsgroup [LinuxOnly]", func() {
+		init()
+		defer cleanup()
+
+		TestPodFsGroupChangePolicyAlways(l.config.Framework, l.config.ClientNodeSelection, l.resource.Pvc)
+	})
+
+	ginkgo.It("pod fsgroup change policy (OnRootMismatch) should change the volume permission to match fsgroup as per root directory ownership [LinuxOnly]", func() {
+		init()
+		defer cleanup()
+
+		TestPodFsGroupChangePolicyOnRootMismatch(l.config.Framework, l.config.ClientNodeSelection, l.resource.Pvc)
+	})
+}
+
+// TestPodFsGroupChangePolicyAlways tests 'Always' policy.
+func TestPodFsGroupChangePolicyAlways(f *framework.Framework, nodeSelection e2epod.NodeSelection, pvc *v1.PersistentVolumeClaim) {
```

Comment:
I think what may be a little hard to understand is that there's actually multiple test cases in this one function. And actually, we want to run the same set of scenarios for both Always and OnRootMistmatch, with the only difference being the expected results. Is it possible to restructure the test cases into each scenario, and then iterate over the two change policies with the expected results?

Reply:
Ack. I have created a table of test cases, with expected results (expected gid), and initial pod conditions.
instead of loop for fsgrppolicy, arranged as a  grouping of tests for each policy. Each test case can be run in parallel.